compiler/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -1972,6 +1975,159 @@ mod tests {
 
         assert_eq!(count_loads(b2, &main.dfg), 1);
         // The repeated loads to v2 should remain as it is passed as a reference argument.
+        assert_eq!(count_loads(b3, &main.dfg), 4);
+    }
+
+    #[test]
+    fn keep_repeat_loads_with_alias_store() {
+        // v13 aliases v11. We want to make sure that a repeat load to v11 with a store
+        // to its alias v13 in between the repeat loads does not remove those loads.
+        //
+        // acir(inline) fn main f0 {
+        //     b0():
+        //       v0 = allocate
+        //       store Field 0 at v0
+        //       v2 = allocate
+        //       store v0 at v2
+        //       jmp b1(Field 0)
+        //     b1(v3: Field):
+        //       v4 = eq v3, Field 0
+        //       jmpif v4 then: b2, else: b3
+        //     b2():
+        //       v5 = load v2
+        //       store Field 2 at v5
+        //       v8 = add v3, Field 1
+        //       jmp b1(v8)
+        //     b3():
+        //       v9 = load v0
+        //       v10 = eq v9, Field 2
+        //       constrain v9 == Field 2
+        //       v11 = load v2
+        //       v13 = load v2
+        //       v14 = load v13
+        //       constrain v14 == Field 2
+        //       jmp b4()
+        //     b4():
+        //       v15 = load v11
+        //       store Field 2 at v13
+        //       v16 = load v11
+        //       v17 = eq v15, Field 2
+        //       constrain v16 == Field 2
+        //       return
```

Comment:
This doesn't quite cover the case in my last comment since from looking at this one block we can find out that v15 and v16 are aliases. We should have block arguments instead that are both aliased to the same value

Reply:
Also, why is the test so long? Do you need more than 2 blocks?